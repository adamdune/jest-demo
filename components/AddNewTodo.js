const { useState } = require("react")
import PropTypes from "prop-types"

const AddNewTodo = ({ createTodo }) => {
  const [newTodo, setNewTodo] = useState("")

  const onClickAddNewTodo = () => {
    if (newTodo) {
      createTodo(newTodo)
      setNewTodo("")
    }
  }

  return (
    <div className="flex flex-col items-center">
      <input
        type="text"
        value={newTodo}
        onChange={(e) => setNewTodo(e.target.value)}
        className="bg-gray-800 p-2 outline-none w-64"
        placeholder="Enter your todo"
      />
      <div className="h-2" />
      <button onClick={onClickAddNewTodo} className="bg-blue-700 px-2 py-1">
        Add New Todo
      </button>
    </div>
  )
}

AddNewTodo.propTypes = {
  createTodo: PropTypes.func.isRequired,
}

export default AddNewTodo
