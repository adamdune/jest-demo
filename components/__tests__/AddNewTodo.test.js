import { fireEvent, render, screen } from "@testing-library/react"
import AddNewTodo from "../AddNewTodo"

describe("AddNewTodo", () => {
  it("when input is populated, the value is reflected ", () => {
    render(<AddNewTodo createTodo={() => {}} />)

    const input = screen.getByPlaceholderText("Enter your todo")

    expect(input).toHaveDisplayValue("")

    fireEvent.change(input, {
      target: { value: "This is a todo" },
    })

    expect(input).toHaveDisplayValue("This is a todo")
  })

  it("while input is filled and button is clicked, onClickAddNewTodo is called and input is emptied", () => {
    const callback = jest.fn()

    render(<AddNewTodo createTodo={callback} />)

    const input = screen.getByPlaceholderText("Enter your todo")

    fireEvent.change(input, {
      target: { value: "This is a todo" },
    })

    fireEvent.click(screen.getByText("Add New Todo"))

    expect(callback).toHaveBeenCalledWith("This is a todo")
    expect(input).toHaveDisplayValue("")
  })

  it("while input is empty and button is clicked, onClickAddNewTodo is not called", () => {
    const callback = jest.fn()

    render(<AddNewTodo createTodo={callback} />)

    const input = screen.getByPlaceholderText("Enter your todo")

    expect(input).toHaveDisplayValue("")

    fireEvent.click(screen.getByText("Add New Todo"))

    expect(callback).toHaveBeenCalledTimes(0)
  })
})
