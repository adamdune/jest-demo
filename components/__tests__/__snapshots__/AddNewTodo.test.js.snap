// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddNewTodo renders correctly 1`] = `
<div
  className="flex flex-col items-center"
>
  <input
    className="bg-gray-800 p-2 outline-none w-64"
    onChange={[Function]}
    placeholder="Enter your todo"
    type="text"
    value=""
  />
  <div
    className="h-2"
  />
  <button
    className="bg-blue-700 px-2 py-1"
    onClick={[Function]}
  >
    Add New Todo
  </button>
</div>
`;
